clippy_lints/src/absolute_symbol_paths.rs

```diff
@@ -0,0 +1,119 @@
+use clippy_utils::diagnostics::span_lint;
+use clippy_utils::is_from_proc_macro;
+use rustc_hir::intravisit::{walk_qpath, Visitor};
+use rustc_hir::{def_id::LOCAL_CRATE, HirId, QPath};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::hir::nested_filter::OnlyBodies;
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::Span;
+use std::iter::once;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for usage of symbols through absolute paths, like `std::env::current_dir`.
+    ///
+    /// ### Why is this bad?
+    /// Many codebases have their own style when it comes to symbol importing, but one that is
+    /// seldom used is using absolute paths *everywhere*. This is generally considered unidiomatic,
+    /// and you should add a `use` statement.
+    ///
+    /// Note: One exception to this is code from macro expansion - this does not lint such cases, as
+    /// using absolute paths is the proper way of referencing symbols in one.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let x = std::f64::consts::PI;
+    /// ```
+    /// Use any of the below instead, or anything else:
+    /// ```rust
+    /// # use std::f64;
+    /// # use std::f64::consts;
+    /// # use std::f64::consts::PI;
+    /// let x = f64::consts::PI;
+    /// let x = consts::PI;
+    /// let x = PI;
+    /// use std::f64::consts as f64_consts;
+    /// let x = f64_consts::PI;
+    /// ```
+    #[clippy::version = "1.72.0"]
+    pub ABSOLUTE_SYMBOL_PATHS,
+    style,
+    "checks for usage of a symbol without a `use` statement"
+}
+impl_lint_pass!(AbsoluteSymbolPaths => [ABSOLUTE_SYMBOL_PATHS]);
+
+pub struct AbsoluteSymbolPaths {
+    pub absolute_symbol_paths_max_segments: u64,
+    pub absolute_symbol_paths_allow_std: bool,
+}
+
+impl LateLintPass<'_> for AbsoluteSymbolPaths {
+    fn check_crate(&mut self, cx: &LateContext<'_>) {
+        let Self {
+            absolute_symbol_paths_max_segments,
+            absolute_symbol_paths_allow_std,
+        } = *self;
+
+        cx.tcx.hir().visit_all_item_likes_in_crate(&mut V {
+            cx,
+            absolute_symbol_paths_max_segments,
+            absolute_symbol_paths_allow_std,
+        });
+    }
```

Comment:
No because we need to have the list of them all by the time we get to expressions, we could use `check_crate_post` for the expressions but then the msrv logic and such won't work

Reply:
Could you elaborate what you mean? I see no lists or MSRV handling in this lint

Reply:
Hmm I must've been thinking of a different lint, yeah looks fine actually. Maybe `get_parent` returns the `QPath` for a `Path`'s id? Shouldn't matter though as the only variant of `QPath` with a `Path` is `Resolved`